web/src/inbox_ui.js

```diff
@@ -0,0 +1,875 @@
+import * as unread from "./unread";
+import * as unread_ops from "./unread_ops";
+import * as unread_ui from "./unread_ui";
+import * as util from "./util";
+
+let pms_dict = {};
+let topics_dict = {};
+let streams_dict = {};
+let include_muted = false;
+let update_triggered_by_user = false;
+
+const COLUMNS = {
+    COLLAPSE_BUTTON: 0,
+    RECIPIENT: 1,
+    UNREAD_COUNT: 2,
+};
+let col_focus = COLUMNS.RECIPIENT;
+let row_focus = 0;
+
+const ls_key = "inbox_mute_filter";
+const ls = localstorage();
+let search_keyword = "";
+const INBOX_SEARCH_ID = "inbox_search";
+const MUTED_FILTER_ID = "include_muted";
+export let current_focus_id = INBOX_SEARCH_ID;
+
+const STREAM_HEADER_PREFIX = "stream_header_";
+const CONVERSATION_ID_PREFIX = "inbox_";
+
+function get_row_from_conversation_key(key) {
+    return $(`#${CONVERSATION_ID_PREFIX}` + CSS.escape(`${key}`));
+}
+
+function save_filter() {
+    ls.set(ls_key, include_muted);
+}
+
+export function show() {
+    if (narrow.has_shown_message_list_view) {
+        narrow.save_pre_narrow_offset_for_reload();
+    }
+
+    if (is_visible()) {
+        return;
+    }
+
+    left_sidebar_navigation_area.narrow_to_inbox();
+    stream_list.handle_narrow_deactivated();
+
+    $("#message_feed_container").hide();
+    $("#inbox_view").show();
+    set_visible(true);
+
+    unread_ui.hide_unread_banner();
+    narrow_state.reset_current_filter();
+    message_view_header.render_title_area();
+    narrow.handle_middle_pane_transition();
+
+    narrow_state.reset_current_filter();
+    narrow.update_narrow_title(narrow_state.filter());
+    message_view_header.render_title_area();
+    pm_list.handle_narrow_deactivated();
+    search.clear_search_form();
+    complete_rerender();
+    compose_closed_ui.set_standard_text_for_reply_button();
+}
+
+export function hide() {
+    const $focused_element = $(document.activeElement);
+
+    if ($("#inbox_view").has($focused_element)) {
+        $focused_element.trigger("blur");
+    }
+
+    $("#message_feed_container").show();
+    $("#inbox_view").hide();
+    set_visible(false);
+
+    // This solves a bug with message_view_header
+    // being broken sometimes when we narrow
+    // to a filter and back to recent topics
+    // before it completely re-rerenders.
+    message_view_header.render_title_area();
+
+    // Fire our custom event
+    $("#message_feed_container").trigger("message_feed_shown");
+
+    // This makes sure user lands on the selected message
+    // and not always at the top of the narrow.
+    navigate.plan_scroll_to_selected();
+}
+
+function get_topic_key(stream_id, topic) {
+    return stream_id + ":" + topic;
+}
+
+function get_stream_key(stream_id) {
+    return "stream_" + stream_id;
+}
+
+function get_stream_container(stream_key) {
+    return $(`#${CSS.escape(stream_key)}`);
+}
+
+function get_topics_container(stream_id) {
+    const $topics_container = get_stream_header_row(stream_id).next();
```

Comment:
Should we pass a class name to `.next`? I always worry with this sort of code that it's fragile in the case that we make some mostly invisible adjustment to the HTML structure.

Reply:
Chnaged the function to use a better selector.

Reply:
```
function get_topics_container(stream_id) {
    const $topics_container = $(
        `#${CSS.escape("stream_" + stream_id)} .inbox-topic-container`,
    ).expectOne();
    return $topics_container;
```

This still seems not ideal? At least that ID of just `stream_{stream_id}` seems likely not globally unique across the app; we have lots of stream lists.

Reply:
Ok, changed back to use get_stream_header_row

Reply:
This seems not actually changed yet.

Reply:
It's changed now.

```js
function get_topics_container(stream_id) {
    const $topics_container = get_stream_header_row(stream_id)
        .next(".inbox-topic-container")
        .expectOne();
    return $topics_container;
}
```